desktopIconsIntegration.js

```diff
@@ -0,0 +1,159 @@
+ * it uses in each side of each monitor.
+ *
+ * DON'T SEND PATCHES TO THIS FILE TO THE EXTENSION MAINTAINER. SEND THEM TO
+ * DESKTOP ICONS NG MAINTAINER: https://gitlab.com/rastersoft/desktop-icons-ng
+ *
+ * In the *enable()* function, create a *DesktopIconsUsableAreaClass()*
+ * object with
+ *
+ *     new DesktopIconsIntegration.DesktopIconsUsableAreaClass(object);
+ *
+ * Now, in the *disable()* function just call to the *destroy()* method before
+ * nullifying the pointer. You must create a new object in enable() the next
+ * time the extension is enabled.
+ *
+ * In your code, every time you change the margins, you should call first to
+ * *resetMargins()* method to clear the current margins, and then call to
+ * *setMargins(...)* method as many times as you need to set the margins in each
+ * monitor. You don't need to call it for all the monitors, only for those where
+ * you are painting something. If you don't set values for a monitor, they will
+ * be considered zero.
+ *
+ * The margins values are relative to the monitor border.
+ *
+ *******************************************************************************/
+
+const GLib = imports.gi.GLib;
+const Main = imports.ui.main;
+
+const ExtensionUtils = imports.misc.extensionUtils;
+const Me = ExtensionUtils.getCurrentExtension();
+
+const IDENTIFIER_UUID = "130cbc66-235c-4bd6-8571-98d2d8bba5e2";
+
+var DesktopIconsUsableAreaClass = class {
+    constructor() {
+        this._extensionManager = Main.extensionManager;
+        this._timedMarginsID = 0;
+        this._margins = {};
+        this._emID = this._extensionManager.connect('extension-state-changed', (_obj, extension) => {
+            if (!extension)
+                return;
+
+            // If an extension is being enabled and lacks the DesktopIconsUsableArea object, we can avoid launching a refresh
+            if (extension.state === ExtensionUtils.ExtensionState.ENABLED) {
+                this._sendMarginsToExtension(extension);
+                return;
+            }
+            // if the extension is being disabled, we must do a full refresh, because if there were other extensions originally
+            // loaded after that extension, those extensions will be disabled and enabled again without notification
+            this._changedMargins();
+        });
+    }
+
+    /**
+     * Sets or updates the top, bottom, left and right margins for a
+     * monitor. Values are measured from the monitor border (and NOT from
+     * the workspace border).
+     *
+     * @param {int} monitor Monitor number to which set the margins.
+     *                      A negative value means "the primary monitor".
+     * @param {int} top Top margin in pixels
+     * @param {int} bottom Bottom margin in pixels
+     * @param {int} left Left margin in pixels
+     * @param {int} right Right margin in pixels
+     */
+    setMargins(monitor, top, bottom, left, right) {
+        this._margins[monitor] = {
```

Comment:
As per https://github.com/micheleg/dash-to-dock/pull/1541#discussion_r790048572, I'd say we should do:

```suggestion
        if (top <= 0 && bottom <= 0 && left  <= 0 && right  <= 0) {
            delete this._margins[monitor];
            this._changedMargins();
         }
        this._margins[monitor] = {
```

Reply:
As I said, I prefer to keep the API as simple as possible. Doing this would mean changing the API itself, so I think it is not a good idea (remember that this file is also used by dash-to-panel and hide-top-bar).

Reply:
I see, would it change the result though?

Reply:
There's still something I'm not sure about this, as with the current `DesktopIconsUsableAreaClass` code in case a single dock is removed (like a disconnected monitor) there would not be a way to notify a single change?

It's true that right now dash-to-dock delete all the docks on monitors-changed, but could actually change in theory (or in future), as it could just add / move docks around. In such case, will it break DesktopIcons?

Reply:
No, there isn't possible to notify a single change. I preferred to keep the interface between extensions as simple as possible, so every time there is a change, the whole current status is passed. Since I wanted to also allow this interface to be used by other extensions, not only by DING (let's say: an extension that allows to paint widgets in the background with CPU/disk,memory/network/whatever usage, could also be interested in know where to not paint them to avoid being occluded by a dock/bar; using the same class than in DING would allow it without having to implement another set of MRs in dash-to-dock, dash-to-panel, hide-top-bar, and any other extension that uses this), so keeping a simple interface was paramount to avoid other programmers to back off.